source/openqasm/parser/antlr/qasm_parser.py

```diff
@@ -0,0 +1,908 @@
+            ),
+            self.visit(ctx.designator()) if ctx.designator() else None,
+        )
+
+    @span
+    def visitBranchingStatement(self, ctx: qasm3Parser.BranchingStatementContext):
+        return BranchingStatement(
+            condition=self.visit(ctx.expression()),
+            if_block=self.visit(ctx.programBlock()[0]),
+            else_block=self.visit(ctx.programBlock()[1]) if ctx.getChildCount() > 5 else None,
+        )
+
+    def visitProgramBlock(self, ctx: qasm3Parser.ProgramBlockContext):
+        if ctx.LBRACE():
+            return [self.visit(statement) for statement in list(ctx.getChildren())[1:-1]]
+        else:
+            return [self.visit(ctx.getChild(0))]
+
+    @span
+    def visitControlDirective(self, ctx: qasm3Parser.ControlDirectiveContext):
+        if ctx.returnStatement():
+            return self.visit(ctx.returnStatement())
+        elif ctx.getChild(0).getText() == "break":
+            return BreakStatement()
+        elif ctx.getChild(0).getText() == "continue":
+            return ContinueStatement()
+        elif ctx.getChild(0).getText() == "end":
+            return EndStatement()
+
+    @span
+    def visitLoopStatement(self, ctx: qasm3Parser.LoopStatementContext):
+        if ctx.loopSignature().getChild(0).getText() == "while":
+            return WhileLoop(
+                while_condition=self.visit(ctx.loopSignature().expression()),
+                block=self.visit(ctx.programBlock()),
+            )
+        else:  # For In Loop
+            return ForInLoop(
+                loop_variable=add_span(
+                    Identifier(ctx.loopSignature().Identifier().getText()),
+                    get_span(ctx.loopSignature().Identifier()),
+                ),
+                set_declaration=self.visit(ctx.loopSignature().setDeclaration()),
+                block=self.visit(ctx.programBlock()),
+            )
+
+    def visitSetDeclaration(self, ctx: qasm3Parser.SetDeclarationContext):
+        if ctx.Identifier():
+            return add_span(
+                Identifier(name=ctx.Identifier().getText()),
+                get_span(ctx),
+            )
+        elif ctx.rangeDefinition():
+            return self.visit(ctx.rangeDefinition())
+        else:
+            return [self.visit(expression) for expression in ctx.expressionList().expression()]
+
+    @span
+    def visitRangeDefinition(self, ctx: qasm3Parser.RangeDefinitionContext):
+        # start, end, are all optional as in [:]
+        start = None
+        end = None
+        step = None
+        colons_seen = 0
+
+        for child in ctx.getChildren():
+            if isinstance(child, qasm3Parser.ExpressionContext):
+                expression = self.visit(child)
+                if colons_seen == 0:
+                    start = expression
+                elif colons_seen == 1:
+                    end = expression
+                else:
+                    step = expression
```

Comment:
I believe that this line should be
```
step = end
end = expression
```
The spec says
> A range is written as a:b or a:c:b where a, b, and c are integers (signed or unsigned).

https://qiskit.github.io/openqasm/language/types.html#register-concatenation-and-slicing

Reply:
Good catch! Fixed.